screwdriver/api/api.go

```diff
@@ -33,102 +38,199 @@ type ResponseError struct {
+			return nil, fmt.Errorf("Screwdriver API Response unparseable: status=%d, err=%v", statusCode, err)
 		}
 		return nil, res
 	case 5:
-		return nil, fmt.Errorf("%v: Screwdriver API has internal server error", res.StatusCode)
+		return nil, fmt.Errorf("Screwdriver API has internal server error: statusCode=%d", statusCode)
 	default:
-		return nil, fmt.Errorf("Unknown error happen while communicate with Screwdriver API")
+		return nil, fmt.Errorf("Unknown error happen while communicate with Screwdriver API: Statuscode=%d", statusCode)
 	}
 }
 
 // GetCommand returns Command from Screwdriver API
-func (c client) GetCommand(smallSpec *Command) (*Command, error) {
-	cmd := new(Command)
-	uri, err := url.Parse(c.baseURL)
+func (c client) GetCommand(smallSpec *util.CommandSpec) (*util.CommandSpec, error) {
+	uri, err := parseURL(c.baseURL)
 	if err != nil {
-		return nil, fmt.Errorf("The base Screwdriver API is invalid %q", c.baseURL)
+		return nil, fmt.Errorf("Failed to parse URL on GET: %v", err)
 	}
 	uri.Path = path.Join(uri.Path, "commands", smallSpec.Namespace, smallSpec.Name, smallSpec.Version)
-	req, err := http.NewRequest("GET", uri.String(), strings.NewReader(""))
-	if err != nil {
-		return nil, fmt.Errorf("Failed to create request about command to Screwdriver API: %v", err)
-	}
 
-	req.Header.Set("Content-type", "application/json")
-	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.jwt))
-	res, err := c.client.Do(req)
+	// Request to api
+	const contentType = "application/json"
+	// No payload
+	payload := bytes.NewBuffer([]byte(""))
+
+	bodyBytes, statusCode, err := c.httpRequest("GET", uri.String(), c.jwt, contentType, payload)
+
 	if err != nil {
-		return nil, fmt.Errorf("Failed to get command from Screwdriver API: %v", err)
+		return nil, fmt.Errorf("Failed to get http response: %v", err)
 	}
-	defer res.Body.Close()
-	body, err := handleResponse(res)
+
+	// Check response body
+	bodyBytes, err = handleResponse(bodyBytes, statusCode)
 
 	if err != nil {
 		return nil, err
 	}
 
-	err = json.Unmarshal(body, cmd)
+	cmd := new(util.CommandSpec)
+	err = json.Unmarshal(bodyBytes, cmd)
 	if err != nil {
 		return nil, fmt.Errorf("Failed to unmarshal command: %v", err)
 	}
 	return cmd, nil
 }
+
+func specToPayloadBytes(commandSpec *util.CommandSpec) (specBytes []byte, err error) {
+	// Generate yaml payload bytes
+	commandSpecBytes, err := json.Marshal(&commandSpec)
+	var payload util.PayloadYaml
+	payload.Yaml = string(commandSpecBytes)
+	specBytes, err = json.Marshal(&payload)
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+func (c client) PostCommand(specPath string, commandSpec *util.CommandSpec) (string, error) {
+	// Generate URL
+	uri, err := parseURL(c.baseURL)
+	if err != nil {
+		return "", fmt.Errorf("Failed to parse URL on POST: %v", err)
+	}
+	uri.Path = path.Join(uri.Path, "commands")
+
+	// Generate payload
+
+	// Prepare body
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
```

Comment:
This function looks bit bigger. I think it is good to divide this function into small functions. For example divide into two function. One create body and the another one do http request.

Reply:
Nice suggestion! Thank you!
I've fixed.